ops/terraform/services/pipeline/modules/bfd_pipeline_scheduler/lambda_src/pipeline_scheduler.py

```diff
@@ -0,0 +1,411 @@
+            elif event_type == S3EventType.OBJECT_REMOVED:
+                # If an object is removed from Incoming, that could mean that the Pipeline loaded
+                # the data load it was from _or_ an operator removed the data load manually. Either
+                # case, we want to check if this load is now completely gone from Incoming and
+                # remove any corresponding scheduled actions if so so the Pipeline doesn't scale
+                # unnecessarily.
+                if not _is_incoming_folder_empty(data_load=data_load):
+                    print(
+                        f"The data load {data_load.name} has not yet been fully removed from"
+                        " Incoming. Exiting..."
+                    )
+                    return
+
+                # The invoking event was for the last file in this load being removed from Incoming;
+                # remove the data load's corresponding scheduled action, if it exists:
+                invalid_scheduled_action_name = (
+                    f"{SCALE_OUT_FUTURE_LOAD_ACTION_NAME_PREFIX}{calendar.timegm(data_load.timestamp.utctimetuple())}"
+                    if is_future_load
+                    else SCALE_OUT_IMMEDIATELY_ACTION_NAME
+                )
+                print(
+                    f"Data load {data_load} has no data in Incoming; it has either been loaded or"
+                    " removed by an external operator. Trying to remove corresponding scheduled"
+                    f" action {invalid_scheduled_action_name}..."
+                )
+                try:
+                    autoscaling_client.delete_scheduled_action(
+                        AutoScalingGroupName=PIPELINE_ASG_NAME,
+                        ScheduledActionName=invalid_scheduled_action_name,
+                    )
+                    print(f"Scheduled action {invalid_scheduled_action_name} successfully deleted")
+                except autoscaling_client.exceptions.ClientError as ex:
+                    print(
+                        "An error occurred when attempting to delete the scheduled action"
+                        f" {invalid_scheduled_action_name}. It is likely that this scheduled action"
+                        f" had already been invoked and the data load {data_load.name} has already"
+                        " been loaded successfully."
+                    )
+        elif (
+            pipeline_data_status == PipelineDataStatus.DONE
+            and event_type == S3EventType.OBJECT_CREATED
+        ):
+            # If the invoking event was from a file being created in Done, we can assume that a
+            # Pipeline Data Load is either in-progress or being completed (as the Pipeline will move
+            # Incoming files to Done when they are finished loading); we only want to schedule the
+            # Pipeline to scale-in when there is no ongoing data load and there are no upcoming
+            # future loads in the _immediate_ future
```

Comment:
Minor clarification that hopefully doesn't impact things; if someone _creates_ a folder and then puts data in it, it will be copied to done but the empty date-stamped folder will stick around until deleted manually. (This sometimes happens if something is being tested manually). I'm assuming theres some AWS funkiness where pipeline doesnt have permissions to delete the manually created folder or something.

Reply:
That's a good point and something I overlooked. It appears that the leftover folder is counted as an object when listing objects with a given prefix.

I'll update `_is_incoming_folder_empty(...)` to check only for valid RIF types and ignore all other S3 objects when determining if the `Incoming` folder is empty.